src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -29309,30 +29325,275 @@ class C<T>
+        [WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
+        public void EnforcedInMethodBody_MaybeNullWhen(string type)
+        {
+            var source = @"
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+class C<T, TClass, TNotNull>
+    where TClass : class
+    where TNotNull : notnull
+{
+    static void GetValue(TYPE x, [MaybeNull] out TYPE y)
+    {
+        y = x;
+    }
+
+    static bool TryGetValue(TYPE x, [MaybeNullWhen(true)] out TYPE y)
+    {
+        y = x;
+        return y == null;
+    }
+
+    static bool TryGetValue2(TYPE x, [MaybeNullWhen(true)] out TYPE y)
+    {
+        y = x;
+        return y != null;
+    }
+
+    static bool TryGetValue3(TYPE x, [MaybeNullWhen(false)] out TYPE y)
+    {
+        y = x;
+        return y == null;
+    }
+
+    static bool TryGetValue4(TYPE x, [MaybeNullWhen(false)] out TYPE y)
+    {
+        y = x;
+        return y != null;
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { MaybeNullAttributeDefinition, MaybeNullWhenAttributeDefinition, source.Replace("TYPE", type) });
+            comp.VerifyDiagnostics();
+        }
+
         [Fact, WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
-        public void EnforcedInMethodBody_Misc()
+        public void EnforcedInMethodBody_MaybeNullWhen_ReferenceType()
+        {
+            var source = @"
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+class C<TClass>
+    where TClass : class
+{
+    static bool TryGetValue(TClass x, [MaybeNullWhen(true)] out TClass y)
+    {
+        y = x;
+        return y == null;
+    }
+
+    static bool TryGetValue2(TClass x, [MaybeNullWhen(true)] out TClass y)
+    {
+        y = x;
+        return y != null; // 1
+    }
+
+    static bool TryGetValue3(TClass x, [MaybeNullWhen(false)] out TClass y)
+    {
+        y = x;
+        return y == null; // 2
+    }
+
+    static bool TryGetValue4(TClass x, [MaybeNullWhen(false)] out TClass y)
+    {
+        y = x;
+        return y != null;
+    }
+}
+";
+            // Note: this is a situation where it is annoying that we're learning from pure null tests
+            var comp = CreateNullableCompilation(new[] { MaybeNullWhenAttributeDefinition, source });
+            comp.VerifyDiagnostics(
+                // (17,9): error CS8762: Parameter 'y' may not have a null value when exiting with 'false'.
+                //         return y != null; // 1
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return y != null;").WithArguments("y", "false").WithLocation(17, 9),
+                // (23,9): error CS8762: Parameter 'y' may not have a null value when exiting with 'true'.
+                //         return y == null; // 2
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return y == null;").WithArguments("y", "true").WithLocation(23, 9)
+                );
+        }
+
+        [Theory]
```

Comment:
Some more cases I'd like to see coverage for:

- Suppression (am I expected to put `!` on the value being assigned? or on the expression being returned?)
- Composition
  - Should be able to use a decorator pattern on methods with equivalent sets of MaybeNullWhen/NotNullWhen attributes
  - Should be able to compose methods with "equivalent" combinations of attributes e.g. a `[NotNullWhen(true)] out string? s1`  composes with a `[MaybeNullWhen(false)] out string s2`

Reply:
In terms of suppression, you can't suppress this new warning directly (except maybe with `#nullable disable warnings` or explicit diagnostic suppression). You can also put non-null values into the parameters in question, with `!` suppressions if needed.

I'll add tests for composition. Could you clarify the decorator scenario you had in mind?

Reply:
➡️ added tests for composition

Reply:
I was thinking of:

```cs
using System.Collections.Generic;

class DictionaryDecorator
{
    readonly Dictionary<string, string> _dict;
    DictionaryDecorator(Dictionary<string, string> dict)
    {
        _dict = dict;
    }

    public bool TryGetValue(string key, [MaybeNullWhen(false)] out string value)
    {
        return _dict.TryGetValue(key, out value);
    }
}
```

which is not all that different from simple method composition, I'll grant :)

---
In reply to: [369371615](https://github.com/dotnet/roslyn/pull/41098#discussion_r369371615) [](ancestors = 369371615)